### PR TITLE
DAOS-12945 rebuild: add DER_VOS_PARTIAL_UPDATE

### PR DESCRIPTION
**Pull request by wangdi1**
DAOS-12945 rebuild: add DER_VOS_PARTIAL_UPDATE
Add DER_VOS_PARTIAL_UPDATE to tell the error of
vos partial update, so rebuild can retry for this
error.

Several DEBUG INFO and tests.

Remove duplicate target check in pl_map_extend(), since
multiple shards are allowed in the same target in 2.4.

Remove some unnecessary parameters in placement since
we do not allow concurrent reintegration and extend.

Required-githooks: true
Allow-unstable-test: true
Test-tag: pr rebuild ec
Test-feature: rebuild ec
Signed-off-by: Di Wang <di.wang@intel.com>

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/include/daos_errno.h [hunk 1]
```diff
@@ -289,7 +289,9 @@ extern "C" {
 	ACTION(DER_NO_CERT,		(DER_ERR_DAOS_BASE + 43),	\
 	       Unable to access one or more certificates)		\
 	ACTION(DER_BAD_CERT,		(DER_ERR_DAOS_BASE + 44),	\
-	       Invalid x509 certificate)
+	       Invalid x509 certificate)				\
+	ACTION(DER_VOS_PARTIAL_UPDATE,	(DER_ERR_DAOS_BASE + 45),	\
+	       VOS partial update error)
```
Comment (jolivier23) [inline src/include/daos_errno.h:294]: Same epoch partial overwrite of VOS array value disallowed

**Review by jolivier23 — APPROVED** [review]

File: src/vos/evtree.c [hunk 2]
```diff
@@ -2721,7 +2721,7 @@ evt_ent_array_fill(struct evt_context *tcx, enum evt_find_opc find_opc,
 						DF_RECT" overlaps with "DF_RECT
 						"\n", DP_RECT(rect),
 						DP_RECT(&rtmp));
-					rc = -DER_NO_PERM;
+					rc = -DER_VOS_PARTIAL_UPDATE;
```
Comment (liuxuezhao) [inline src/vos/evtree.c:2724]: now for failed rebuild the half-rebuilt data will be discarded, so how could the vos partial update hit again?

**Review by liuxuezhao — DISMISSED** [review]

**Comment by wangdi1** [comment]
failure due to DAOS-12883

File: src/tests/ftest/control/dmg_pool_query_ranks.py [hunk 3]
```diff
@@ -141,8 +143,21 @@ def test_pool_query_ranks_mgmt(self):
...
+            while counter < 3:
+                try:
+                    result = self.pool.reintegrate(rank)
+                except CommandFailure:
+                    self.log.debug("dmg command failed retry")
+                else:
+                    cmd_succeed = True
+                    break
+                counter += 1
+                time.sleep(3)
```
Comment (daltonbohning) [inline src/tests/ftest/control/dmg_pool_query_ranks.py:158]: Suggested simplification
```suggestion
            cmd_succeed = False
            for _ in range(3):
                try:
                    result = self.pool.reintegrate(rank)
                    cmd_succeed = True
                    break
                except CommandFailure:
                    self.log.debug("dmg command failed retry")
                time.sleep(3)
```

File: src/tests/ftest/control/dmg_pool_query_ranks.py [hunk 4]
```diff
@@ -141,8 +143,21 @@ def test_pool_query_ranks_mgmt(self):
...
+
+            self.assertTrue(cmd_succeed, result)
```
Comment (daltonbohning) [inline src/tests/ftest/control/dmg_pool_query_ranks.py:160]: We should print a more descriptive failure message. For example
```suggestion
            self.assertTrue(cmd_succeed, "pool reintegrate failed: {}".format(result))
```

File: src/tests/ftest/daos_test/suite.yaml [hunk 5]
```diff
@@ -23,7 +23,7 @@ timeouts:
   test_daos_epoch_recovery: 104
   test_daos_md_replication: 104
   test_daos_rebuild_simple: 1800
-  test_daos_drain_simple: 500
+  test_daos_drain_simple: 2000
```
Comment (daltonbohning) [inline src/tests/ftest/daos_test/suite.yaml:26]: Do we really need another 25 minutes for this test?

**Review by liuxuezhao — APPROVED** [review]

File: src/vos/evtree.c (hunk 2 above)
Comment (wangdi1) [reply src/vos/evtree.c:2724]: there are not enough log for me to understand this yet, but "retry" can fix this this. I will collect more log to understand this.

My suspicion is due to enumeration limit.

**Pull request merged**